### ISSUE TITLE
Unit conversion functions

### ISSUE THREAD

**Issue by ctf0** [date-fns/date-fns#732]
is it possible to have like a helper function that can convert a date to another ?

ex.
```js
convert(2, 'hrs', 'milli')
convert(1, 'month', 'hrs')
etc...
```

**Comment by edorivai** [comment]
What is it exactly that you want to accomplish? In case you want to convert durations, this can be done with math!

```js
// 2 hours in milliseconds
2 * 60 * 60 * 1000; // 7200000

// 1 month in hours (approximately, because it depends on your definition of the length of a month)
const averageDaysInMonth = 365.25 / 12;
1 * averageDaysInMonth * 24; // 730.5
```

**Comment by ctf0** [comment]
exactly that but for ppl that suck in math like myself, the **helper** will the perfect answer

**Comment by kossnocorp** [comment]
This is a good idea, however, I don't like the proposed API. I suggest making it explicit, like `secondsToMinutes(45)`, `hoursToDays(123)`, etc.

PRs are welcome.

**Comment by edorivai** [comment]
Oh my, should this be a cross map of all different time units? Also, we should probably make a table of definitions. Like how many days is a month? A year? Do we do gap years?

**Comment by Lakston** [comment]
I can try do do something for this request but we would have to define the API and what conversions we would plan to support because some can get a bit tricky as mentioned by @edorivai , if you want to convert minutes to months, would you take the current month length ? allow for a month argument (that would not fit with a simple, explicit API) ?

The easiest ones to start with would be : 

- seconds => minutes
- seconds => hours
- seconds => days
- minutes => seconds
- minutes => hours
- minutes => days

When getting into months / years territory it gets more complicated with months duration and gap years.

Maybe an optional argument such as `minutesToYears(15962, true)` where `true` defines if the year is a gap year or not and `secondsToMonths(15472, 30)` where `30` would be an optional month duration.

**Comment by ctf0** [comment]
imho using the explicit api will be hard to maintain while using the one from the main topic is easier as it can be done using switches.

on -a-side-note apparently this is hard to create because of the indifferences in month days & year months

**Comment by leshakoss** [comment]
For the record, without any ambiguity it is possible to convert between these groups of units:
- milliseconds ↔ seconds ↔ minutes ↔ hours
- days ↔ weeks
- months ↔ quarters ↔ years

**Comment by edorivai** [comment]
@leshakoss hours :left_right_arrow: days?

**Comment by edorivai** [comment]
Damn, you're right

**Comment by kossnocorp** [comment]
@ctf0

> imho using the explicit api will be hard to maintain while using the one from the main topic is easier as it can be done using switches.

I can't see how switches could be easier to maintain than dedicated functions.

> on -a-side-note apparently this is hard to create because of the indifferences in month days & year months

This is yet another reason why the API should be explicit. If we'd allow converting anything to anything it would result in issues that are hard to debug.

Besides that, implicit API would be inconsistent with the rest of the library.

---

As @leshakoss noted, we're looking for explicit API that allows converting only compatible units:

> - milliseconds ↔ seconds ↔ minutes ↔ hours
> - days ↔ weeks
> - months ↔ quarters ↔ years

I like the `unitToUnit` naming scheme, i.e. `daysToWeeks(12)` cc @Lakston 

Another question is what should we return floats or integers. @leshakoss WDYT?

**Comment by edorivai** [comment]
@kossnocorp If we say we only want to convert compatible units, is it then really worth it to increase the API surface? I mean, the compatible conversions are arguably trivial to do by hand.

I could see the use in date-fns actually defining "sensible defaults" for the problematic conversions, but then again, maybe in this case there is no such thing as sensible defaults.

If I would personally have to make the call I would probably say it's not worth the extra API surface. But if enough people think it's worth it, then be my guest and add it.

**Comment by ctf0** [comment]
@kossnocorp what i meant by a switch is to have a fluent method, like for example
u search google on how to convert 1gb to mb, but now we want to also know what is 1gb in kb, so instead of re-searching u just select the new type you want to convert to.

so in our care you just write one method, and just change the types to whatever you want without having to memorize/search all the api methods to convert from one thing to another.

**Comment by Lakston** [comment]
> If I would personally have to make the call I would probably say it's not worth the extra API surface. But if enough people think it's worth it, then be my guest and add it.

I would agree with that, expanding the API with functions that could be done that easily doens't sound great. 

I proposed my help because I was waiting for the issues about the localisations updates were not created yet (I started working on updating the 'fr' locale to match the latest release but had some questions so I was looking for an issue in the mean time)

**Comment by kakadiadarpan** [comment]
If anyone is still looking for a utility like this, check this out: https://github.com/zeit/ms

**Comment by jatin33** [comment]
Can I work on this issue?

**Comment by leshakoss** [comment]
@jatin33 sure!

**Comment by LucasHFS** [comment]
Hi guys I would like to work on these funcitions as my first PR to the lib. What would be the response of these functions, integers or floats?

**Comment by fturmel** [comment]
@tan75 I think this issue can be closed since #2433 was merged.
